Signal/src/ViewControllers/ConversationView/Cells/OWSQuotedMessageView.m

```diff
@@ -0,0 +1,384 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSQuotedMessageView.h"
+#import "ConversationViewItem.h"
+#import "Environment.h"
+#import "OWSBubbleStrokeView.h"
+#import "OWSMessageCell.h"
+#import "Signal-Swift.h"
+#import <SignalMessaging/OWSContactsManager.h>
+#import <SignalMessaging/SignalMessaging-Swift.h>
+#import <SignalMessaging/UIColor+OWS.h>
+#import <SignalMessaging/UIView+OWS.h>
+#import <SignalServiceKit/TSAttachmentStream.h>
+#import <SignalServiceKit/TSMessage.h>
+#import <SignalServiceKit/TSQuotedMessage.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface OWSQuotedMessageView ()
+
+@property (nonatomic, readonly) TSQuotedMessage *quotedMessage;
+@property (nonatomic, nullable, readonly) DisplayableText *displayableQuotedText;
+
+@property (nonatomic, readonly) UIFont *textMessageFont;
+
+@property (nonatomic, nullable) OWSBubbleStrokeView *boundsStrokeView;
+
+@end
+
+#pragma mark -
+
+@implementation OWSQuotedMessageView
+
++ (OWSQuotedMessageView *)quotedMessageViewForConversation:(TSQuotedMessage *)quotedMessage
+                                     displayableQuotedText:(nullable DisplayableText *)displayableQuotedText
+{
+    OWSAssert(quotedMessage);
+
+    return
+        [[OWSQuotedMessageView alloc] initWithQuotedMessage:quotedMessage displayableQuotedText:displayableQuotedText];
+}
+
++ (OWSQuotedMessageView *)quotedMessageViewForPreview:(TSQuotedMessage *)quotedMessage
+{
+    OWSAssert(quotedMessage);
+
+    DisplayableText *_Nullable displayableQuotedText = nil;
+    if (quotedMessage.body.length > 0) {
+        displayableQuotedText = [DisplayableText displayableText:quotedMessage.body];
+    }
+
+    return
+        [[OWSQuotedMessageView alloc] initWithQuotedMessage:quotedMessage displayableQuotedText:displayableQuotedText];
+}
+
+- (instancetype)initWithQuotedMessage:(TSQuotedMessage *)quotedMessage
+                displayableQuotedText:(nullable DisplayableText *)displayableQuotedText
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(quotedMessage);
+
+    _quotedMessage = quotedMessage;
+    _displayableQuotedText = displayableQuotedText;
+    _textMessageFont = OWSMessageCell.defaultTextMessageFont;
+
+    return self;
+}
+
+- (BOOL)hasQuotedAttachment
+{
+    return (self.quotedMessage.contentType.length > 0
+        && ![NSObject isNullableObject:self.quotedMessage.contentType equalTo:OWSMimeTypeOversizeTextMessage]);
+}
+
+- (BOOL)hasQuotedAttachmentThumbnail
+{
+    return (self.quotedMessage.contentType.length > 0
+        && ![NSObject isNullableObject:self.quotedMessage.contentType equalTo:OWSMimeTypeOversizeTextMessage] &&
```

Comment:
Can this check be rolled into the hasThumbnailForMimeType? Presumably there would never be a thumbnail for the oversized text content type.

Reply:
Those methods have slightly different semantics.  

* `[TSAttachment hasThumbnailForMimeType]` indicates whether or not the attachment expects to be able to generate a _thumbnail image_ for the give content type.
* This "quoted reply" view shows a "attachment view" for all attachments _except_ oversize text attachments.

Regardless, this got reworked in the next PR.  LEt's revisit there.  I think we can clarify by renaming this method.